pkgs/swift2objc/lib/src/transformer/transform.dart

```diff
@@ -8,24 +8,55 @@ import '../ast/_core/interfaces/nestable_declaration.dart';
 import '../ast/declarations/compounds/class_declaration.dart';
 import '../ast/declarations/compounds/struct_declaration.dart';
 import '../ast/declarations/globals/globals.dart';
+import '_core/dependencies.dart';
 import '_core/unique_namer.dart';
 import 'transformers/transform_compound.dart';
 import 'transformers/transform_globals.dart';
 
 typedef TransformationMap = Map<Declaration, Declaration>;
 
-List<Declaration> transform(List<Declaration> declarations) {
+Set<Declaration> generateDependencies(Iterable<Declaration> decls, {Iterable<Declaration>? allDecls}) {
+  final dependencies = <Declaration>{};
+  final dependencyVisitor = DependencyVisitor();
+
+  var _d = decls;
+
+  while (true) {
+    final deps = _d.fold<Set<String>>(
```

Comment:
This algorithm is pretty hard to follow. A simpler approach would be to give `DependencyVisitor` a member `Set` that it outputs all the dependencies into. `generateDependencies` would just construct a `DependencyVisitor`, then iterate over every element of `decls` and call `dependencyVisitor.visitDeclaration`, then return the `Set` that `dependencyVisitor` constructed.

`DependencyVisitor.visitDeclaration` would just DFS from the given element to any child element not in that `Set`. This would also mean you don't have to pass around the `context` variable through all those `visit` methods in `DependencyVisitor`, because the `context` would be replaced by this member `Set`.

Reply:
The reason for recursiveness is because of at the moment, the algorithm generates bindings for all parts of all dependencies of filtered types, and since these parts (e.g methods and properties) may also require types as well, those are generated as well recursively.

What can be done, is to pass the declarations to `DependencyVisitor`, allow it to perform the recursive visiting, and then return the output.

Reply:
yeah, all that recursion should happen inside the visitor